Flask server code review

Meteorstrike1 opened the pull request:
- Unit testing
- Reconfiguring of create app set up
- Updating content tables and column names to lowercase

RolfdeSterke commented on flask-server/introverseSQL.sql:
```diff
@@ -61,39 +61,39 @@ VALUES
-    Book_Script VARCHAR(1000),
-    Book_Image VARCHAR(100) UNIQUE
+CREATE TABLE books ( 
+	book_id INTEGER PRIMARY KEY NOT NULL,
```
I'm not a huge fan of prefixing the attribute names with the entity name. It should be somewhat obvious that this id refers to the book it's part of. 

I would reserve this notation for when creating foreign keys. E.g. if there is a reference to the author, that could be the author_id.

Meteorstrike1 replied:
Yeah that's fair, makes a lot of sense. I'll move away from doing that in the future and practice with some more SQL projects.

RolfdeSterke commented on flask-server/migrations/versions/e11b182cd2e2_updated_table_and_column_names_for_.py:
```diff
@@ -0,0 +1,154 @@
+"""Updated table and column names for content tables to be lowercase
```
Very diligent to make these!
In a production environment it's very handy to have the last upgraden en downgrade at hand for releasing and potential rollbacks if needed.

RolfdeSterke commented on flask-server/models/forum_models.py:
```diff
@@ -0,0 +1,50 @@
+
+    Attributes
+    ----------
+    post_id : int
+        id of the message
+    post_content : str (text)
+        text content of the message
+    post_category : str
+        category of the message
+    post_author : str
+        author of the message
+    post_date : datetime
+        time and date that the message was posted
```
Being nitpicky on these. It's good that you made these, but in their state they are kind of... obvious. What would be interesting to note here is for example that the id is system generated or not. And if it's used as a PK or not.
If there are any unique constraints on them etc.

Meteorstrike1 replied:
Updated descriptions
[47059db](https://github.com/Angel2001-programmer/Introverse/pull/16/commits/47059dbe82c7564b29c46e503189daaa6c0c9356)

RolfdeSterke commented on flask-server/routes/content.py:
```diff
@@ -2,41 +2,40 @@
 })
 
 
 @content_ns.route("/books")
```
For this file it might be interesting to look into query parameters. 
These will definitely work, but it might be interesting to search for Thrillers of a certain author as well.

Meteorstrike1 replied:
Will look into mulitple query parameters in the future, that is a nice idea!

RolfdeSterke commented on flask-server/routes/content.py:
```diff
@@ -164,32 +163,40 @@ def get(self, id):
         return game
+
+
+@content_ns.route("/hello")
```
Hello!

RolfdeSterke commented on flask-server/tests/test_forum_api.py:
```diff
@@ -0,0 +1,128 @@
+
+    def test_create_post_successful(self):
+        """Test creating a post, login required for @jwt_required route"""
+        register_response = self.client.post("/user/register",
+            json = {
+                "username": "testuser",
+                "first_name": "test",
+                "last_name": "user",
+                "email": "testuser@test.com",
+                "password": "mytestpassword"
+            }
+        )
```
I can imagine there is a "beforeEach" like function with pytesting too. It'd be code that is executed before every single test. THat would be useful to move this user creation to, since otherwise you will be repeating this code a lot.

If thats not possible, you could make a create_user() function to do this and call that (or like a setup() function) at the start of the relevant tests, just to clean this up a bit. 

Once you get more complex cases these can get very finicky.

Meteorstrike1 replied:
Yeah that would be very useful, I'll look into.

Meteorstrike1 replied:
Ah I think the beforeEach equivalent is setUp, and the end of test equivalent is tearDown. For mine I have the table creation. Think you can add whatever you like but to test every different scenario probably better I don't default mock sign up members. So being smarter with mocking or seeing if there are any tools unittest has for it would be the way forward I think.

Meteorstrike1 replied:
To reduce repetition I have made a function in the main test API file to create a default user and updated the existing tests to use it
75faf62
Will add more for other cases that have duplication

Meteorstrike1 replied:
Testing "complete"
56f808b
Rough summary of tests and associated functions attached
![image](https://github.com/Angel2001-programmer/Introverse/assets/130995689/222d9123-3f6a-4fd0-b2a7-b4620d301611)

RolfdeSterke commented on flask-server/tests/test_user_api.py:
```diff
@@ -0,0 +1,267 @@
+        self.assertEqual(status_code, 201)
+
+    def test_register_successful_max_length(self):
+        """Test successful registration of user, username, first name and last name max characters"""
```
Part of your registration logic is to login the user and return it's token right?
It would be nice to fetch that token here as well and validate that it's filled and works for authentication.

Meteorstrike1 replied:
Will look into this more and look into different methods of authentication as well for something more secure.

Meteorstrike1 commented on flask-server/create_db.py:
```diff
@@ -1,22 +1,27 @@
-for db in my_cursor:
-    print(db)
\ No newline at end of file
+def create_db_if_not_exist(dbname):
```
If they already exist it will print that they exist. Maybe I should rename the function? "Check if exist and create if not" or something.
![image](https://github.com/Angel2001-programmer/Introverse/assets/130995689/cf8ab1a9-b1f0-4208-8b84-2407e820784c)

Meteorstrike1 replied:
Have renamed function to make its purpose more clear

Meteorstrike1 commented:
Have addressed some of the issues and also completed testing and tweaking the routes with error handling. Happy with it as a complete Flask server for now.
